lib/class-wp-rest-sidebars-controller.php

```diff
@@ -116,129 +117,71 @@ public function permissions_check() {
-					}
-				}
-
-				if ( $registered_widget_id ) {
-					// Old-style widget.
-					$update_control = $wp_registered_widget_updates[ $registered_widget_id ];
-					$_POST          = wp_slash( $input_widget['settings'] );
-					call_user_func( $update_control['callback'] );
-				}
-			}
-			ob_end_clean();
+	public function get_items( $request ) {
+		$data = array();
+		foreach ( (array) wp_get_sidebars_widgets() as $id => $widgets ) {
+			list( , $sidebar ) = $this->get_sidebar( $id );
 
-			$sidebar_widgets_ids[] = $input_widget['id'];
+			$data[] = $this->prepare_response_for_collection(
+				$this->prepare_item_for_response( $sidebar, $request )
+			);
 		}
 
-		// Update sidebar to only consist of the widgets we just processed.
-		$sidebars                = wp_get_sidebars_widgets();
-		$sidebars[ $sidebar_id ] = $sidebar_widgets_ids;
-		wp_set_sidebars_widgets( $sidebars );
-
-		$request = new WP_REST_Request( 'GET' );
-		$request->set_param( 'id', $sidebar_id );
-
-		return $this->get_item( $request );
+		return rest_ensure_response( $data );
 	}
 
 	/**
-	 * Returns a list of sidebars (active or inactive)
+	 * Returns the given sidebar
 	 *
 	 * @param WP_REST_Request $request The request instance.
 	 *
-	 * @return WP_REST_Response
-	 * @global array $wp_registered_sidebars
+	 * @return WP_REST_Response|WP_Error
 	 */
-	public function get_items( $request ) {
-		$data = array();
-		foreach ( (array) wp_get_sidebars_widgets() as $id => $widgets ) {
-			$sidebar = self::get_sidebar( $id )[1];
+	public function get_item( $request ) {
+		list( $exists, $sidebar ) = $this->get_sidebar( $request['id'] );
 
-			$data[] = $this->prepare_item_for_response( $sidebar, $request )->get_data();
+		if ( ! $exists && 'wp_inactive_widgets' !== $request['id'] ) {
+			return new WP_Error( 'rest_sidebar_not_found', __( 'No sidebar exists with that id.', 'gutenberg' ), array( 'status' => 404 ) );
 		}
 
-		return rest_ensure_response( $data );
+		return $this->prepare_item_for_response( $sidebar, $request );
 	}
 
 	/**
-	 * Returns the given sidebar
+	 * Updates the sidebar.
 	 *
 	 * @param WP_REST_Request $request The request instance.
 	 *
 	 * @return WP_REST_Response
 	 */
-	public function get_item( $request ) {
-		$sidebar = self::get_sidebar( $request['id'] )[1];
+	public function update_item( $request ) {
+		if ( isset( $request['widgets'] ) ) {
+			$sidebars = wp_get_sidebars_widgets();
+
+			foreach ( $sidebars as $sidebar_id => $widgets ) {
+				foreach ( $widgets as $i => $widget_id ) {
+					if ( $sidebar_id !== $request['id'] && in_array( $widget_id, $request['widgets'], true ) ) {
+						unset( $sidebars[ $sidebar_id ][ $i ] );
+						$sidebars['wp_inactive_widgets'][] = $widget_id;
+					}
+				}
+			}
+
+			$sidebars[ $request['id'] ] = $request['widgets'];
+
+			wp_set_sidebars_widgets( $sidebars );
+		}
 
+		$request['context'] = 'edit';
+
+		list( , $sidebar ) = $this->get_sidebar( $request['id'] );
```

Comment:
What is happening on this line?

Reply:
The actual sidebar details are returned in the second parameter of the array. I don't really love the `get_sidebar` method, but we can change that after this get's merged since it won't change the output.